platform/pulp/plugin/changeset/main.py

```diff
@@ -0,0 +1,200 @@
+        Associate content.
+
+        Args:
+            content (RemoteContent): The content to be associated.
+        """
+        log.debug(_('Associate content.'))
+        try:
+            association = RepositoryContent(
+                repository=self.importer.repository,
+                content=content)
+            association.save()
+        except IntegrityError:
+            # Duplicate
+            pass
+
+    def _disassociate(self, content):
+        """
+        (Un)associate content with the repository.
+
+        Args:
+            content (pulp.plugin.Content): A content model to be unassociated.
+        """
+        log.debug(_('Disassociate content.'))
+        RepositoryContent.objects.delete(
+            repository=self.importer.repository,
+            content=content)
+        for artifact in content.artifacts:
+            DownloadCatalog.objects.delete(
+                importer=self.importer,
+                artifact__in=artifact)
+
+    def _add_content(self):
+        """
+        Add wanted content.
+
+        Yields:
+            AddReport: A report for each content added.
+        """
+        activity = _('Adding content')
+        log.info(activity)
+        with Batch(DownloadIterator(self.adds, self.deferred)) as batch:
+            with ProgressBar(message=activity, total=len(self.adds)) as bar:
+                for plan in batch():
+                    artifact = plan.download.attachment
+                    artifact.model.downloaded = True
+                    content = artifact.content
+                    if not plan.succeeded:
+                        task = Task()
+                        task.append_non_fatal_error(plan.error)
+                        bar.increment()
+                        report = ChangeReport(ChangeReport.ADDED, content.model)
+                        report.errors.append(plan.error)
+                        yield report
+                        continue
+                    content.settle()
+                    if not content.settled:
+                        continue
+                    bar.increment()
+                    with transaction.atomic():
+                        self._update_catalog(content)
+                        self._associate(content.model)
+                    report = ChangeReport(ChangeReport.ADDED, content.model)
+                    yield report
+
+    def _update_catalog(self, content):
+        """
+        Update the content catalog.
+        Add (or update) an entry for each artifact.
+
+        Args:
+            content (RemoteContent): A newly added content.
+        """
+        for artifact in content.artifacts:
+            log.info(_('Cataloging: {a}').format(a=self))
+            try:
+                entry = DownloadCatalog()
+                entry.importer = self.importer
+                entry.url = artifact.download.url
+                entry.artifact = artifact.model
+                entry.save()
+            except IntegrityError:
```

Comment:
I expect mirror lists will be handled by a custom `Download` object.  The `DownloadCatalog` needs to have the uniqueness constraint added.